R/mod_common_forms.R

```diff
@@ -93,60 +93,236 @@ mod_common_forms_server <- function(
+        session$userData$update_checkboxes[[form]] <- NULL
+        session$userData$review_records[[form]] <-
+          dplyr::rows_upsert(
+            session$userData$review_records[[form]],
+            input[[review_selection]][c("id", "reviewed")],
+            by = "id"
+          ) |> 
+          dplyr::filter(!is.na(reviewed)) |> 
+          # Ensure that only the current subject is being reviewed
+          dplyr::semi_join(
+            subset(r$review_data, subject_id == r$subject_id & item_group == form),
+            by = "id"
           ) |> 
-          dplyr::select(-dplyr::starts_with("SAE"))
+          # Only update records where the review status is being changed
+          dplyr::anti_join(
+            subset(r$review_data, subject_id == r$subject_id & item_group == form),
+            by = c("id", "reviewed")
+          ) |> 
+          dplyr::arrange(id)
+        
+        
+        # Update the table's data reactive
+        df <- do.call(tbl_data, list())
+        
+        update_row <- dplyr::distinct(input[[review_selection]], reviewed, row_id)
+        row_ids <- df$o_reviewed |> lapply(\(x) x[["row_id"]]) |> unlist()
+        df[row_ids == update_row$row_id, "o_reviewed"] <- list(list(
+          modifyList(df[row_ids == update_row$row_id,]$o_reviewed[[1]], 
+                     list(updated = switch(update_row$reviewed, "Yes" = TRUE, "No" = FALSE, NA)))
+        ))
+        do.call(tbl_data, list(df))
+      })
+    })
+    
+    # Any time the data in the form table is updated, "show all" is toggled,
+    # or the subject being viewed is changed, the server data for the datatable
+    # needs to be updated
+    observe({
+      req(!is.null(input$show_all_data))
+      req(common_form_data())
+      DT::dataTableAjax(common_form_proxy$session, 
+                      subset(common_form_data(), input$show_all_data | subject_id == r$subject_id),
+                      rownames = FALSE,
+                      outputId = common_form_proxy$rawId)
+      req(SAE_data())
+      DT::dataTableAjax(SAE_proxy$session, 
+                      subset(SAE_data(), input$show_all_data | subject_id == r$subject_id), 
+                      rownames = FALSE,
+                      outputId = SAE_proxy$rawId)
+    })
+    # Any time the review table is updated, "show all" is toggled, or the
+    # subject being viewed is changed, the datatable should be reloaded to show
+    # the new data
+    observeEvent(reload_data(), {
+      req(!is.null(input$show_all_data))
+      req(common_form_data())
+      DT::reloadData(common_form_proxy)
+      req(SAE_data())
+      DT::reloadData(SAE_proxy)
+    }, ignoreInit = TRUE)
+    
+    observeEvent(r$subject_id, {
+      req(common_form_data())
+      reload_data(reload_data() + 1)
+      df <- common_form_data() |> 
+        dplyr::mutate(o_reviewed = Map(\(x, y) modifyList(x, list(updated = NULL, disabled = y)), o_reviewed, subject_id != r$subject_id))
```

Comment:
Does the `disabled` variable currently do anything? Or does it not yet have a function?

Reply:
The `disabled` variable was added to handle when all subjects are being viewed. This value is used in the datatable object. The checkbox gets rendered as disabled if you aren't currently viewing that subject.